docs/designdoc/allow_customize_containers.md

```diff
@@ -0,0 +1,105 @@
+# Allow users to customize containers
+
+## Context
+
+Currently, MOCO has containers that are automatically added by the system in addition to containers added by the user.
+(e.g. `agent`, `moco-init` etc...)
+Containers automatically added by the system do not allow user customization.
+
+There are cases where users want to customize container resources, as in Issue [cybozu-go/moco#235](https://github.com/cybozu-go/moco/issues/235).
+This design document examines how users can customize containers that are automatically added by the system.
+
+## Goals
+
+* No breaking changes
+* Allows users to customize containers that are automatically added by the system
+* Extendable functionality for future customization beyond container resources
+* Users can only customize authorized fields
+  * Do not allow customization of fields that would make operation untenable (e.g. `command`)
+
+## Non-goals
+
+* No customization provided at the time of implementation other than requested container resources
+* No enhancement will be added to the `moco.cybozu.com/v1beta1` API.
+
+## ActualDesign
+
+Add `.spec.podTemplate.overwriteContainers` field to MySQLCluster.
+
+```yaml
+spec:
+  podTemplate:
+    spec:
+      containers:
+      - name: mysqld
+        image: quay.io/cybozu/mysql:8.0.28
+    overwriteContainers:
+    - name: agent
+      resources:
+        requests:
+          cpu: 50m
+```
+
+Pseudo-code:
+
+```go
+type PodTemplateSpec struct {
+	ObjectMeta
+	Spec                PodSpecApplyConfiguration
+	OverwriteContainers []OverwriteContainer // added
+}
+
+type OverwriteContainer struct {
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:Enum=agent;moco-init;slow-log;mysqld-exporter
+	Name      string
```

Comment:
Actually, this should be defined as a dedicated type like:

```go
// +kubebuilder:validation:Enum=agent;moco-init;slow-log;mysqld-exporter
type ContainerName string

const (
    AgentName ContainerName = "agent"
    ...
)
```

Also, this definition should replace the previous definition in `pkg/constants`.
https://github.com/cybozu-go/moco/blob/main/pkg/constants/container.go#L3

Please replace the references of the previous definition with this one.

Reply:
@ymmt2005 
Thanks for the comments!
Since the container names defined in pkg/constants are not all overwritable containers (I'm thinking that the mysqld container is user-specifiable, so there is no need to define it as an overwritable container) I was thinking of adding the following definition to the v1beta2 package, what do you think?

```go
type PodTemplateSpec struct {
	// Standard object's metadata.  The name in this metadata is ignored.
	// +optional
	ObjectMeta `json:"metadata,omitempty"`

	// Specification of the desired behavior of the pod.
	// The name of the MySQL server container in this spec must be `mysqld`.
	Spec PodSpecApplyConfiguration `json:"spec"`

	// OverwriteContainers overwrites the container definitions provided by default by the system.
	// +optional
	OverwriteContainers []OverwriteContainer `json:"overwriteContainers,omitempty"`
}

// +kubebuilder:validation:Enum=agent;moco-init;slow-log;mysqld-exporter
type OverwriteableContainerName string

const (
	AgentContainerName             OverwriteableContainerName = constants.AgentContainerName
	InitContainerName              OverwriteableContainerName = constants.InitContainerName
	SlowQueryLogAgentContainerName OverwriteableContainerName = constants.SlowQueryLogAgentContainerName
	ExporterContainerName          OverwriteableContainerName = constants.ExporterContainerName
)

type OverwriteContainer struct {
	// +kubebuilder:validation:Required
	Name OverwriteableContainerName `json:"name"`

	// +optional
	Resources *ResourceRequirementsApplyConfiguration `json:"resources,omitempty"`
}
```

Reply:
Looks good!

Reply:
Updated: https://github.com/cybozu-go/moco/pull/394/commits/befa8c856c659e77952180891251b55153ee6b6a